libnmstate/appliers/bond.py

```diff
@@ -126,3 +129,36 @@ def fix_bond_option_arp_monitor(cur_iface_state):
         and "arp_ip_target" not in bond_options
     ):
         bond_options["arp_ip_target"] = ""
+
+
+def discard_merged_data_on_mode_change(
+    merged_iface_state, desire_iface_state, current_iface_state
+):
+    """
+    When bond mode changed, use original desire bond options instead of merging
+    from current state.
+    """
+    if (
+        desire_iface_state
+        and current_iface_state
+        and desire_iface_state.get(Bond.CONFIG_SUBTREE, {}).get(Bond.MODE)
+        and current_iface_state.get(Bond.CONFIG_SUBTREE, {}).get(Bond.MODE)
+        and desire_iface_state[Bond.CONFIG_SUBTREE][Bond.MODE]
+        != current_iface_state[Bond.CONFIG_SUBTREE][Bond.MODE]
+    ):
+        merged_bond_options = merged_iface_state[Bond.CONFIG_SUBTREE].get(
+            Bond.OPTIONS_SUBTREE
+        )
+        desire_bond_options = desire_iface_state[Bond.CONFIG_SUBTREE].get(
+            Bond.OPTIONS_SUBTREE
+        )
+        if merged_bond_options != desire_bond_options:
+            logging.warning(
+                "Discarding all current bond options as bond mode changed"
+            )
+            merged_iface_state[BOND_MODE_CHANGED_METADATA] = True
```

Comment:
This is the 4th functionality added to this function.
(Check that he mode has changed, check that the options changed, overwrite the options and now add metadata info)

Binding them together here gives little chance for these to be used in other steps. For example, the metadata part is more naturally fitting the metadata step (under the metadata module).

What do you think about splitting and moving the marking to the metadata module?

Reply:
Changed to 3 functions:
 * `generate_bond_mode_change_metadata()`, used to create metadata on bond mode change.
 * `discard_merged_data_on_mode_change()`, used after merge to discard merged data.
 * `is_bond_mode_changed()`, used by NM applier to deactivate the interface.